src/themis/secure_message.c

```diff
@@ -77,6 +83,212 @@ themis_status_t themis_gen_ec_key_pair(uint8_t* private_key,
+    return THEMIS_KEY_RSA_PRIVATE;
+  }
+  if(!memcmp(key, RSA_PUB_KEY_PREF, strlen(RSA_PUB_KEY_PREF))){
+    return THEMIS_KEY_RSA_PUBLIC;
+  }
+  if(!memcmp(key, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF))){
+    return THEMIS_KEY_EC_PRIVATE;
+  }
+  if(!memcmp(key, EC_PUB_KEY_PREF, strlen(EC_PUB_KEY_PREF))){
+    return THEMIS_KEY_EC_PUBLIC;
+  }
+
+  return THEMIS_KEY_INVALID;
+}
+
+themis_status_t themis_is_valid_key(const uint8_t* key, size_t length){
+  const soter_container_hdr_t* container=(const void*)key;
+
+  if(!key || (length<sizeof(soter_container_hdr_t))){
+    return THEMIS_INVALID_PARAMETER;
+  }
+  if(length!=ntohl(container->size)){
+    return THEMIS_INVALID_PARAMETER;
+  }
+  if(SOTER_SUCCESS!=soter_verify_container_checksum(container)){
+    return THEMIS_DATA_CORRUPT;
+  }
+
+  return THEMIS_SUCCESS;
+}
+
+static bool valid_private_key(const uint8_t* private_key, size_t private_key_length){
+  if(themis_is_valid_key(private_key, private_key_length)==THEMIS_SUCCESS){
+    themis_key_kind_t private_key_kind=themis_get_key_kind(private_key, private_key_length);
+    switch(private_key_kind){
+      case THEMIS_KEY_EC_PRIVATE:
+      case THEMIS_KEY_RSA_PRIVATE:
+        return true;
+      default:
+        break;
+    }
+  }
+  return false;
+}
+
+static bool valid_public_key(const uint8_t* public_key, size_t public_key_length){
+  if(themis_is_valid_key(public_key, public_key_length)==THEMIS_SUCCESS){
+    themis_key_kind_t public_key_kind=themis_get_key_kind(public_key, public_key_length);
+    switch(public_key_kind){
+      case THEMIS_KEY_EC_PUBLIC:
+      case THEMIS_KEY_RSA_PUBLIC:
+        return true;
+      default:
+        break;
+    }
+  }
+  return false;
+}
+
+static bool matching_key_kinds(const uint8_t* private_key, size_t private_key_length,
+                               const uint8_t* public_key, size_t public_key_length)
+{
+  themis_key_kind_t private_key_kind=themis_get_key_kind(private_key, private_key_length);
+  themis_key_kind_t public_key_kind=themis_get_key_kind(public_key, public_key_length);
+  if(private_key_kind==THEMIS_KEY_EC_PRIVATE && public_key_kind==THEMIS_KEY_EC_PUBLIC){
+    return true;
+  }
+  if(private_key_kind==THEMIS_KEY_RSA_PRIVATE && public_key_kind==THEMIS_KEY_RSA_PUBLIC){
+    return true;
+  }
+  return false;
+}
+
+themis_status_t themis_secure_message_encrypt(const uint8_t* private_key,
+                                              const size_t private_key_length,
+                                              const uint8_t* public_key,
+                                              const size_t public_key_length,
+                                              const uint8_t* message,
+                                              const size_t message_length,
+                                              uint8_t* encrypted_message,
+                                              size_t* encrypted_message_length)
+{
+  THEMIS_CHECK_PARAM(private_key!=NULL);
```

Comment:
can we remove these checks and leave the only `themis_is_valid_key` call for private/public keys?

Reply:
Indeed, null and size checks are redundant if `themis_is_valid_key` is used. Nice catch.